### DISCUSSION TITLE
How can I get requiv@primary and its errors？

### DISCUSSION THREAD

**Discussion by mzb123** [phoebe-project/phoebe2#658] [General]
Hi，
In the contact model, I want to get the value of requiv@primary and its error, but it always tells me 'potential probably out of bounds for contact envelope'. However, I found that the potential is between pot_min and pot_max. I would like to know how I can get the requiv@primary and its error.
Also I added
b.flip_constraint('pot', solve_for='requiv@primary')
b.flip_constraint('fillout_factor', solve_for='pot')
Thanks
![requiv](https://user-images.githubusercontent.com/118281186/201942408-7f83892e-56aa-465d-bdb0-c2ac6f0e0a71.png)

**Comment by kecnry** [comment]
This is an interesting situation I haven't seen before.  For some reason, the conversion between fillout factor and radius is failing... that could happen if at least one of the samples is numerically very close to the limits but could also be a bug somewhere in the conversion logic since the exception raising the error you're seeing is a catch-all.  To try to track that down... can you enable a debugging logger (`phoebe.logger('debug')`) immediately before running the line that fails and send us the last ~10 lines?

**Reply by mzb123** [reply]
Thank you for your answer. Sorry, I just started to learn about phoebe. Although I have enabled the debug logger (phoebe. logger ('debug ')), I don't know how to view the logger. In addition, if I can know how to calculate this step in phobe, I can also calculate it manually.

**Reply by kecnry** [reply]
By default, it should print output to the screen (terminal or notebook cell if you're in jupyter).  If you want to export to a file instead, use `phoebe.logger(flevel='debug', filename='myfile.txt')`.  Note though that you should only create a single logger per-runtime (calling `phoebe.logger` again may not edit the existing logger).

If you'd rather calculate manually, you would need to propagate all the sample values of fillout-factor (and any other relevant parameter) through the equations to radius and then determine the central value and uncertainties from there.

**Reply by mzb123** [reply]
Sorry, I still haven't been able to solve this problem. I can get the correct result with the following command. b.uncertainties_from_distribution_collection(solution='emcee_sol', parameters=['pot','fillout_factor'], tex=True)
However, for 'requiv@primary' or 'requiv@primary', an error is reported, here is the full error.

TypeError                                 Traceback (most recent call last)
File F:\conda\envs\name_of_phoebe_environment\lib\site-packages\phoebe\constraints\builtin.py:303, in pot_to_requiv_contact(pot, q, sma, compno)
    302 logger.debug("libphoebe.roche_contact_partial_area_volume(nekmin={}, q={}, d={}, pot={}, compno={})".format(nekmin, q, d, pot, compno-1))
--> 303 volume_equiv = libphoebe.roche_contact_partial_area_volume(nekmin, q, d, pot, compno-1, lvolume=True, ldvolume=False, larea=False)['lvolume']
    304 # returns normalized vequiv, should multiply by sma back for requiv in SI

TypeError: roche_contact_partial_area_volume::Problem reading arguments

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
Input In [36], in <cell line: 1>()
----> 1 b.uncertainties_from_distribution_collection(solution='emcee_sol', parameters=['requiv@primary'], tex=True)

File F:\conda\envs\name_of_phoebe_environment\lib\site-packages\phoebe\frontend\bundle.py:9266, in Bundle.uncertainties_from_distribution_collection(self, twig, parameters, sigma, tex, **kwargs)
   9188 def uncertainties_from_distribution_collection(self, twig=None,
   9189                                                parameters=None,
   9190                                                sigma=1, tex=False,
   9191                                                 **kwargs):
   9193     """
   9194     Get (asymmetric) uncertainties for all parameters in a distribution collection
   9195     by first sampling the underlying distribution object(s) 1 million times.
   (...)
   9264     * (distl.Latex object): with methods as_string and as_latex
   9265     """
-> 9266     return self.get_distribution_collection(twig=twig, parameters=parameters, set_labels=kwargs.pop('set_labels', True), **kwargs)[0].uncertainties(sigma=sigma, tex=tex)

File F:\conda\envs\name_of_phoebe_environment\lib\site-packages\phoebe\dependencies\distl\distl.py:4212, in DistributionCollection.uncertainties(self, sigma, tex)
   4180 def uncertainties(self, sigma=1, tex=False):
   4181     """
   4182     Expose (asymmetric or symmetric) uncertainties for the distribution(s) at a given
   4183     value of `sigma`.
   (...)
   4209 
   4210     """
-> 4212     ret_ = [dist.uncertainties(sigma=sigma, tex=tex) for dist in self.dists]
   4214     if tex:
   4215         return Latex.from_list(ret_)

File F:\conda\envs\name_of_phoebe_environment\lib\site-packages\phoebe\dependencies\distl\distl.py:4212, in <listcomp>(.0)
   4180 def uncertainties(self, sigma=1, tex=False):
   4181     """
   4182     Expose (asymmetric or symmetric) uncertainties for the distribution(s) at a given
   4183     value of `sigma`.
   (...)
   4209 
   4210     """
-> 4212     ret_ = [dist.uncertainties(sigma=sigma, tex=tex) for dist in self.dists]
   4214     if tex:
   4215         return Latex.from_list(ret_)

File F:\conda\envs\name_of_phoebe_environment\lib\site-packages\phoebe\dependencies\distl\distl.py:2697, in BaseUnivariateDistribution.uncertainties(self, sigma, tex, unit)
   2672 """
   2673 Expose (asymmetric) uncertainties for the distribution(s) at a given
   2674 value of `sigma`.
   (...)
   2694 
   2695 """
   2696 quantiles = _norm.cdf([-sigma, 0, sigma])
-> 2697 qs = self._return_with_units(self.ppf(quantiles, wrap_at=False), unit=unit, as_quantity=False)
   2699 if tex:
   2700     return _format_uncertainties_asymmetric([self.label],
   2701                                             [self.label_latex],
   2702                                             [unit if unit is not None else self.unit],
   2703                                             [qs])

File F:\conda\envs\name_of_phoebe_environment\lib\site-packages\phoebe\dependencies\distl\distl.py:2262, in BaseUnivariateDistribution.ppf(self, q, unit, as_quantity, wrap_at)
   2226 """
   2227 Expose the percent point function (ppf; iverse of cdf - percentiles) at
   2228 values of `q`.
   (...)
   2259 * (float or array) ppf values of the same type/shape as `q`
   2260 """
   2261 try:
-> 2262     ppf = self.dist_constructor_object.ppf(q)
   2263 except AttributeError:
   2264     raise NotImplementedError("{} does not support ppf".format(self.__class__.__name__))

File F:\conda\envs\name_of_phoebe_environment\lib\site-packages\phoebe\dependencies\distl\distl.py:1082, in BaseDistribution.dist_constructor_object(self)
   1068 """
   1069 Return the instantiated underlying distribution constructor (often the
   1070 scipy.stats random variable object).
   (...)
   1079 * object
   1080 """
   1081 if self._dist_constructor_object_cache is None:
-> 1082     self._dist_constructor_object_cache = self.dist_constructor_func(*self.dist_constructor_args)
   1084 return self._dist_constructor_object_cache

File F:\conda\envs\name_of_phoebe_environment\lib\site-packages\phoebe\dependencies\distl\distl.py:5238, in Function.dist_constructor_args(self)
   5236 @property
   5237 def dist_constructor_args(self):
-> 5238     return self.to_histogram(self.hist_samples, bins=100, wrap_at=False).dist_constructor_args

File F:\conda\envs\name_of_phoebe_environment\lib\site-packages\phoebe\dependencies\distl\distl.py:2822, in BaseUnivariateDistribution.to_histogram(self, N, bins, range, wrap_at)
   2799 def to_histogram(self, N=100000, bins=10, range=None, wrap_at=None):
   2800     """
   2801     Convert the <<class>> distribution to a <Histogram> distribution.
   2802 
   (...)
   2820     * a <Histogram> object
   2821     """
-> 2822     return Histogram.from_data(self.sample(size=N, wrap_at=False, cache_sample=False),
   2823                                bins=bins, range=range,
   2824                                unit=self.unit,
   2825                                label=self.label, label_latex=self._label_latex,
   2826                                wrap_at=wrap_at if wrap_at is not None else self.wrap_at)

File F:\conda\envs\name_of_phoebe_environment\lib\site-packages\phoebe\dependencies\distl\distl.py:5331, in Function.sample(self, size, unit, as_quantity, wrap_at, seed, cache_sample)
   5293 def sample(self, size=None, unit=None, as_quantity=False, wrap_at=None, seed={}, cache_sample=True):
   5294     """
   5295     Sample from the <Function> distribution.
   5296 
   (...)
   5329         shape defined by `size`.
   5330     """
-> 5331     sample = self._sample_from_children(self.func, self.args, self.kwargs, self.vectorized, size=size, seed=seed, cache_sample=cache_sample)
   5333     if isinstance(sample, _units.Quantity):
   5334         if sample.unit is None or sample.unit in [_units.dimensionless_unscaled]:

File F:\conda\envs\name_of_phoebe_environment\lib\site-packages\phoebe\dependencies\distl\distl.py:5290, in Function._sample_from_children(self, func, args, kwargs, vectorized, seed, size, cache_sample, skip_fail)
   5286         raise
   5287 else:
   5288     # NOTE: we have to increment the incoming seed otherwise each sample in the list comprehension will be the same
   5289     # unfortunately this will break any expected correlation if this FunctionDistribution is nested in any parent DistributionCollection, Fuction, or Composite distribution
-> 5290     return _np.asarray([func(*argsi, **{k:v[i] for k,v in kwargs.items()}) for i, argsi in enumerate(args.T)])

File F:\conda\envs\name_of_phoebe_environment\lib\site-packages\phoebe\dependencies\distl\distl.py:5290, in <listcomp>(.0)
   5286         raise
   5287 else:
   5288     # NOTE: we have to increment the incoming seed otherwise each sample in the list comprehension will be the same
   5289     # unfortunately this will break any expected correlation if this FunctionDistribution is nested in any parent DistributionCollection, Fuction, or Composite distribution
-> 5290     return _np.asarray([func(*argsi, **{k:v[i] for k,v in kwargs.items()}) for i, argsi in enumerate(args.T)])

File F:\conda\envs\name_of_phoebe_environment\lib\site-packages\phoebe\constraints\builtin.py:308, in pot_to_requiv_contact(pot, q, sma, compno)
    305     return sma * (3./4 * 1./np.pi * volume_equiv)**(1./3)
    306 except:
    307     # replace this with actual check in the beginning or before function call
--> 308     raise ValueError('potential probably out of bounds for contact envelope')

ValueError: potential probably out of bounds for contact envelope

**Reply by kecnry** [reply]
Were you able to enable the debugging logger?  That would be very helpful in order for us to reproduce the exact parameter values that are triggering the error.  Alternatively, please save the bundle and attach it (either here or you can email it to me directly if you don't want it to be publicly available).

**Reply by kecnry** [reply]
Thanks for sending the files.  This should now be fixed in PHOEBE 2.4.6 (just released).
